app/models/labware_creators/donor_pooling_plate.rb

```diff
@@ -0,0 +1,225 @@
+        request_hash(source_well, dest_plate, additional_parameters)
+      end
+    end
+
+    # Returns the well from a plate that matches the specified location. If no
+    # well is found at the specified location, nil is returned. Used by the
+    # 'request_hash' method to find the destination well.
+    #
+    # @param plate [Sequencescape::Api::V2::Plate] The destination plate.
+    # @param well_location [String] The location to match.
+    # @return [Sequencescape::Api::V2::Well, nil] The well at the specified location,
+    #   or nil if no well is found.
+    def get_well_for_plate_location(plate, well_location)
+      plate.wells.detect { |well| well.location == well_location }
+    end
+
+    # Generates a hash representing a transfer request from a source well to a
+    # destination well. Additional parameters generated by the well filter are
+    # merged into the request hash, i.e.'outer_request' and 'submission_id'.
+    # tag_depth is added to the aliquot attributes.
+    #
+    # @param source_well [Sequencescape::Api::V2::Well] The source well.
+    # @param dest_plate [Sequencescape::Api::V2::Plate] The destination plate.
+    # @param additional_parameters [Hash] Additional parameters to include.
+    # @return [Hash] A hash representing a transfer request.
+    def request_hash(source_well, dest_plate, additional_parameters)
+      dest_location = transfer_hash[source_well][:dest_locn]
+      {
+        'source_asset' => source_well.uuid,
+        'target_asset' => get_well_for_plate_location(dest_plate, dest_location)&.uuid,
+        :aliquot_attributes => {
+          'tag_depth' => tag_depth(source_well)
+        }
+      }.merge(additional_parameters)
+    end
+
+    # Returns a mapping between each source well to a destination location.
+    #
+    # @return [Hash] A hash where each key is a source well and each value is a
+    #   hash with a single key-value pair: { dest_locn: destination_location }.
+    def transfer_hash
+      @transfer_hash ||=
+        pools
+          .each_with_index
+          .with_object({}) do |(pool, index), result|
+            dest_location = WellHelpers.well_at_column_index(index) # column order, 96 wells
+            pool.each { |source_well| result[source_well] = { dest_locn: dest_location } }
+          end
+    end
+
+    # Returns the tag depth for the given source well. The tag depth is the
+    # position of the well in its pool. It is used used as an aliquot attribute
+    # in the transfer request. It is recorded in Sequencescape to avoid tag
+    # clashes.
+    #
+    # @param source_well [Well] The source well for which to retrieve the tag depth.
+    # @return [String] The tag depth as a string, or nil if the well is not in a pool.
+    def tag_depth(source_well)
+      pools.each do |pool|
+        return (pool.index(source_well) + 1).to_s if pool.index(source_well)
+        # index + 1 incase of 0th index
+      end
+    end
```

Comment:
I expect it doesn't make a big difference performance-wise, but seems a little inefficient looping through all the pools here, to find the source well. We could store a hash that could help.

Reply:
That is right. I have removed this method. We are now building a tag_depth_hash with pre-computed tag depths and using the cached result in the request_hash method.